workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using Unity.Entities;
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    /// <summary>
+    /// Retrieves Reader and Writer fields from MonoBehaviours and handles injection into them.
+    /// </summary>
+    public class SpatialOSBehaviourLibrary {
+        // Stores maps from Spatial component IDs to MemberAdapters to reflectively set Readers/Writers for each SpatialOSBehaviour.
+        private readonly Dictionary<Type, Dictionary<uint, IMemberAdapter>> adapterCache
+            = new Dictionary<Type, Dictionary<uint, IMemberAdapter>>();
+
+        // List of component IDs for which a reader/writer is required for each SpatialOSBehaviour
+        private readonly Dictionary<Type, List<uint>> componentReaderIdsForBehaviours = new Dictionary<Type, List<uint>>();
+        private readonly Dictionary<Type, List<uint>> componentWriterIdsForBehaviours = new Dictionary<Type, List<uint>>();
+
+        // Helper objects for querying members with the [Require] attribute
+        private readonly MemberReflectionUtil reflectionUtil = new MemberReflectionUtil(typeof(RequireAttribute));
+
+        private readonly HashSet<Type> badTypes = new HashSet<Type>();
+
+        private readonly ILogDispatcher logger;
+        private const string LoggerName = "SpatialOSBehaviourLibrary";
+
+        private const string BadRequiredMemberWarning
+            = "[Require] attribute found on member that is not Reader or Writer or incorrectly generated, ignoring!";
```

Comment:
Is this ignoring the invalid object or the monobehaviour?

Reply:
Read through the code below. Looks like we ignore the whole monobehaviour (which seems reasonable). Please add that to the message.

Reply:
Actually no, we only ignore the whole monobehaviour if it requires several readers/writers to the same component type. If it requires something nonsensical, we just ignore that member. I'll clarify that.

Reply:
That way round doesn't make sense. If it requires two readers to the same type that should be perfectly well defined. We could warn a user if we think there is literally no use case for it, but it's asking for a well defined thing.
If a monobehaviour requires something that isn't a reader or writer that is an error and means either they meant to use a different attribute or that they expect different behaviour. In that case we should insist on fixing the monobehaviour before running it.